### PR TITLE
:zap: Dévoilé. Cancel. Voix.

### PR DESCRIPTION
**Pull request by floppyzedolfin**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
+build:
+	go build -o bin/littre.out src/cli/main.go
+
+rmgob:
+	rm bin/dict.gob
+
+run-word:
 	go run src/cli/main.go
+
+word:
```
Comment (alienorlatour) [inline Makefile:16]: depends on build?

File: bin/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# bin
```
Comment (floppyzedolfin) [inline bin/README.md]: The only reason for this file's existence is to make sure the `bin/` dir is properly created everywhere. But that's already achieved with the `.gitignore` file. We can discard this one.

File: parsers/gob.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	"os"
+
+	"github.com/ablqk/littre-bot/src/dictionary"
+)
+
+var (
+	gobFile = "bin/dict.gob"
+)
+
+func ParseGob() ([]dictionary.Entry, error) {
```
Comment (floppyzedolfin) [inline parsers/gob.go]: UT it (maybe a single test, covering save + parse, to avoid storing binary files on github)

File: parsers/xml2ram.go [hunk 4]
```diff
@@ -28,14 +29,19 @@ func ParseAlphabet(folderPath string) ([]dictionary.Entry, error) {
...
 	}
 	wg.Wait()
 	if wgErr != nil {
 		return nil, wgErr
 	}
+
+	// save it, for future reference
+	err := saveGob(dico)
+	if err != nil {
+		return nil, fmt.Errorf("unable to save gob: %w", err)
```
Comment (floppyzedolfin) [inline parsers/xml2ram.go]: should merely be a warning

File: parsers/xml2ram.go [hunk 5]
```diff
@@ -70,9 +76,20 @@ func format(x xmlittre) []dictionary.Entry {
 				Term:      e.Terme,
 				Header:    e.Entete.Text,
 				Body: dictionary.EntryBody{
-					Def: v.Num + ". " + v.Text,
+					Def: func()string{if len(v.Num) > 0 {
+						return v.Num + ". " + strings.TrimSuffix(v.Text, "\n")
+					} else {
+						return strings.TrimSuffix(v.Text,"\n")
```
Comment (floppyzedolfin) [inline parsers/xml2ram.go]: the .TrimSuffix didn't really change a lot of things

File: parsers/xml2ram.go [hunk 6]
```diff
@@ -70,9 +76,20 @@ func format(x xmlittre) []dictionary.Entry {
...
+					}}(),
 				},
 			}
+			ee.Body.Quotes = func() []string {
+				quotes := make([]string, 0)
+				for _, q := range v.Cit {
```
Comment (floppyzedolfin) [inline parsers/xml2ram.go:84]: Maybe `'` ?

File: src/cli/main.go [hunk 7]
```diff
@@ -11,25 +11,33 @@ import (
...
 
 func main() {
-	fromXML, err := parsers.ParseAlphabet("parsers/xmlittre-data")
-	if err != nil {
-		fmt.Println(err)
-		return
+	var parsedEntries []dictionary.Entry
+	var err error
+
+	if _, err = os.Stat("bin/dict.gob"); os.IsNotExist(err) {
```
Comment (floppyzedolfin) [inline src/cli/main.go]: defer the knowledge of the gob path to the gob lib

goblin

File: parsers/gob.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	f, err := os.Open(gobFile)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open gob: %w", err)
+	}
+	defer f.Close()
+	decoder := gob.NewDecoder(f)
+	dict := make([]dictionary.Entry, 0)
+	err = decoder.Decode(&dict)
```
Comment (alienorlatour) [inline parsers/gob.go:19]: Does this line work if your slice is not initialised?

File: parsers/gob.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"encoding/gob"
+	"fmt"
+	"os"
+
+	"github.com/ablqk/littre-bot/src/dictionary"
+)
+
+var (
+	gobFile = "bin/dict.gob"
```
Comment (alienorlatour) [inline parsers/gob.go]: This should probably be part of the caller's package, right?

File: parsers/xml2ram.go (hunk 4 above)
Comment (alienorlatour) [inline parsers/xml2ram.go]: I think that an error here should be returned but not block the rest of the process.

File: parsers/xml2ram.go [hunk 10]
```diff
@@ -70,9 +76,20 @@ func format(x xmlittre) []dictionary.Entry {
 				Term:      e.Terme,
 				Header:    e.Entete.Text,
 				Body: dictionary.EntryBody{
-					Def: v.Num + ". " + v.Text,
+					Def: func()string{if len(v.Num) > 0 {
```
Comment (alienorlatour) [inline parsers/xml2ram.go]: I don't like the formatting but I like the idea!

**Review by alienorlatour — CHANGES_REQUESTED** [review]

File: src/cli/main.go (hunk 7 above)
Comment (alienorlatour) [reply src/cli/main.go]: Quite the opposite. The lib should not decide where it writes, if you want it to be reusable.

File: parsers/gob.go (hunk 3 above)
Comment (alienorlatour) [reply parsers/gob.go]: More importantly, you want to test that these two work together. If you change the binary format, you don't want to update references.

Comment (floppyzedolfin) [reply parsers/gob.go]: but a .gob file could have been written by a previous run. If we change the parsing after the writing of the file, we'll face sever issues

File: Makefile (hunk 1 above)
Comment (floppyzedolfin) [reply Makefile:16]: yes, but I didn't want to have `build` relaunched everytime I wanted a word

File: parsers/gob.go (hunk 8 above)
Comment (floppyzedolfin) [reply parsers/gob.go:19]: It _seems_ so: https://golang.org/pkg/encoding/gob/#example__encodeDecode

File: Makefile (hunk 1 above)
Comment (alienorlatour) [reply Makefile:16]: We need to rename `build` then. Call it `bin.out` or something.

File: parsers/gob.go (hunk 8 above)
Comment (alienorlatour) [reply parsers/gob.go:19]: Then l.22 should just be a var declaration.

**Pull request merged**